src/core/datastore/deltaset.js

```diff
@@ -0,0 +1,159 @@
+        .equalTo('collectionName', collectionNameWithTag)
+        .and()
+        .equalTo('query', serializedQuery);
+      return offlineRepo.read(queryCacheCollectionName, queryCacheQuery)
+        .then((deltaSetQueryDocs = []) => {
+          if (deltaSetQueryDocs.length > 0) {
+            return deltaSetQueryDocs[0];
+          }
+
+          return {
+            _id: generateEntityId(),
+            collectionName: collectionNameWithTag,
+            query: serializedQuery
+          };
+        });
+    });
+}
+
+function updateDeltaSetQuery(deltaSetQuery, response) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      deltaSetQuery.lastRequest = response.headers.get(xKivneyRequestStartHeader);
+      return offlineRepo.update(queryCacheCollectionName, deltaSetQuery);
+    });
+}
+
+function readDocs(collectionNameWithTag, query) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      return offlineRepo.read(collectionNameWithTag, query);
+    });
+}
+
+function deleteDocs(collectionNameWithTag, deleted = []) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      if (isArray(deleted) && deleted.length > 0) {
+        const deletedIds = deleted.map((doc) => doc._id);
+        const deleteQuery = new Query().contains('_id', deletedIds);
+        return offlineRepo.delete(collectionNameWithTag, deleteQuery)
+      }
+
+      return deleted;
+    });
+}
+
+function updateDocs(collectionNameWithTag, changed = []) {
+  return repositoryProvider.getOfflineRepository()
+    .then((offlineRepo) => {
+      if (isArray(changed) && changed.length > 0) {
+        return offlineRepo.update(collectionNameWithTag, changed)
+      }
+
+      return changed;
+    });
+}
+
+function makeDeltaSetRequest(collectionNameWithTag, deltaSetQuery, query, options) {
+  const collectionName = stripTagFromCollectionName(collectionNameWithTag);
+  const client = Client.sharedInstance();
+  const request = new KinveyRequest({
+    authType: AuthType.Default,
+    method: RequestMethod.GET,
+    url: format({
+      protocol: client.apiProtocol,
+      host: client.apiHost,
+      pathname: buildCollectionUrl(collectionName, '_deltaset'),
+      query: { since: deltaSetQuery.lastRequest }
+    }),
+    query,
+    timeout: options.timeout,
+    followRedirect: options.followRedirect,
+    cache: options.cache,
+    properties: options.properties,
+    skipBL: options.skipBL,
+    trace: options.trace,
+    client
+  });
+
+  return request.execute()
+    .then((response) => {
+      return updateDeltaSetQuery(deltaSetQuery, response).then(() => response.data);
+    })
+    .then((data) => {
+      return deleteDocs(collectionNameWithTag, data.deleted)
+        .then(() => updateDocs(collectionNameWithTag, data.changed))
+        .then(() => data);
+    });
+}
+
+function makeRequest(collectionNameWithTag, deltaSetQuery, query, options) {
```

Comment:
This function can very easily be split into two - it does two (and a half) things - performs a delta set request, or a regular get request (and the "half" is managing how local store should be updated). It's clearly visible where one, self-sufficient part ends (the first `return`). This should be a separate function.

The name "makeRequest" even suggests things should be different - if you can't think of a descriptive name for a function, it means it's not clear what it does. Which means it should change.

I think the way to solve this is split `makeRequest` into all of its small operations and then "merge" `makeRequest` with `deltaSet`. The resulting function would delegate all operations to other functions and handle common context. This is readable and each separate function would be cohesive.

After "a request" is made, both "paths" (which should be functions) call `updateDeltaSetQuery()`. This would be solved by the above paragraph too.